test/rapi/test_rc_api_runtime.c

```diff
@@ -107,6 +107,42 @@ static void test_init_fetch_game_data_request() {
   rc_api_destroy_request(&request);
 }
 
+static void test_init_fetch_game_data_request_game_hash() {
+  rc_api_fetch_game_data_request_t fetch_game_data_request;
+  rc_api_request_t request;
+
+  memset(&fetch_game_data_request, 0, sizeof(fetch_game_data_request));
+  fetch_game_data_request.username = "Username";
+  fetch_game_data_request.api_token = "API_TOKEN";
+  fetch_game_data_request.game_id = 1234;
+  fetch_game_data_request.game_hash = "ABCDEF0123456789";
+
+  ASSERT_NUM_EQUALS(rc_api_init_fetch_game_data_request(&request, &fetch_game_data_request), RC_OK);
+  ASSERT_STR_EQUALS(request.url, DOREQUEST_URL);
+  ASSERT_STR_EQUALS(request.post_data, "r=patch&u=Username&t=API_TOKEN&g=1234&m=ABCDEF0123456789");
```

Comment:
Are `g` and `m` mutually exclusive? 

What will the server return if `m` isn't a known hash for game `g`?

Will the server changes allow fetching game data without a `g` parameter? (see https://github.com/RetroAchievements/RAWeb/issues/618)

Reply:
It probably makes sense for `m` to always take precedence over `g` if `m` is provided.

In https://github.com/RetroAchievements/RAWeb/blob/40cc59a25a99c0b5abb82bc140cd5c4b8a15240f/app/Platform/Actions/ResolveAchievementSetsAction.php, the game ID for each set is also returned. This is achieved via the `attachCoreGameIds()` function:

![Screenshot 2024-11-13 at 7 15 41 PM](https://github.com/user-attachments/assets/905bd923-5895-4f54-a8f6-ce0d3810c569)

Note in the screenshot above, I've attached a subset to game ID 1 which in the real world has no place being there.

Given this, maybe it makes sense to remove `g` from the patch API call long-term? My gut feeling is it shouldn't be removed until multiset is actually generally available to the public, that way we still have game ID to fall back to use legacy logic if something goes wrong.

Reply:
Server-side, we can't realistically ever remove it since we have no clear path to deprecate old clients.

Client side, I'd prefer to keep it as well, as some clients may not be using rc_client yet, and removing the field would be a breaking change.

I was thinking something more along these lines:
```
    if (api_params->game_hash && *api_params->game_hash)
      rc_url_builder_append_str_param(&builder, "m", api_params->game_hash);
    else
      rc_url_builder_append_unum_param(&builder, "g", api_params->game_id);
```

Reply:
That proposed change definitely makes sense and I'd be happy to do it. The only lingering question in the back of my mind is what if the server for some reason does not respect the `m` param, ie:

* The code to do so is not yet deployed.
* The code to do so is deployed but, for some reason, disabled.

I'm wondering if we may always want to keep `g` on to serve as a fallback value.

Reply:
Generally, when I make API changes, I develop the client and server code at the same time to ensure it's functioning the way I want it to, then open the server PR, and only open the client PR after the server code is merged under the assumption that it will be deployed before the client PR gets merged.

Reply:
It sounds like it may make sense to close this PR until the work in RAWeb is complete, and only re-open it once that work is merged.